### ISSUE TITLE
What are `Context` and `ClientSession` exactly? 

### ISSUE THREAD

**Issue by TK-21st** [jupyterlab/jupyterlab#4636]
Couple of technical questions that I've been avoiding by blindly following examples in codebase but would really like to understand. Here's what I've gathered so far.

## `ClientSession` and `Session`
`Session` constructor has `name`, `path`, `type`.  `ClientSession` has an additional `manager` of type `SessionManager` required, which can be passed in runtime by a handler to `ServiceManager.session`. (see my other issue #4614 ).

### Path
By following example `Notebook`'s source code, `path` is often a unique ID (most likely the _path_ of a notebook (.ipynb) relative to current working directory. When `Session` is launched with a given `path`, if the file with this `path` exists, it will be shown as active (green dot), otherwise nothing changes in filebrowser but the _Running_ tab will show your active session.

### Name 
Often defaults to the same as the `Path` but can be set to be different. The `Name` is what is shown in the _Running_ tab so that if your `path` is `notebook.ipynb` but `name` is `hello`, then the running session reflected in the left _Running_ tab will show `hello`.

### Type (really have not idea what this does)
For notebook it defaults to `File`, tried changing it but doesn't seem to do anything.

## `Context`
I've seen this mostly referenced to by `Document` (see [readthedocs](https://jupyterlab.readthedocs.io/en/stable/developer/documents.html?highlight=document)). It contains the actual `ClientSession` associated with the document (not just the interface but the actual session object). 

## Questions
1. What does `ClientSession`'s `Name` affect other than what is displayed in _Running_ ? Should I even change it to something different from `Path` ?
2. What does `ClientSession`'s `Type` do???
3. What really is a `context?` When would you use `Context` instead of just `ClientSession` (maybe use both document's `model` and `session` as shown in [imageviewer](https://github.com/jupyterlab/jupyterlab/blob/8de8cf6d361329fe975627d09a0cc12a517eec4a/packages/imageviewer/src/widget.ts#L39))

**Comment by ian-r-rose** [comment]
Hi @TK-21st, I think you are mostly on the right track here. The `ClientSession` is meant to be a more user-friendly wrapper around the `Session` object, so they should have mostly the same functionality. `ClientSession` is meant to be run in the browser (so it has access to the DOM), whereas `Session` can be run on Node.

The session path is used in a few places throughout the code base to identify whether a kernel is associated with a given document. I think you are right that `name` is used mostly for display purposes (such as in the running tab). I am not really sure what the `type` field is intended for, I don't see it used much in the code base.

The `Context` is the manages interface between the in-memory document model and the server. As such, it manages both the file-system interactions and the kernel interactions (hence, it has both a `path` and a `session` reference). If the document has no kernel (e.g., an image file), then the `session` reference should be `null`.

To summarize, and address your specific questions:
1. I think the `name` is used for display purposes. It could be rendered in other places than just the `Running` tab, but that might be the only place at the moment.
2. Not sure. It doesn't seem to be used much.
3. The `Context` and `ClientSession` do very different things. All documents that are backed by a file have a `context` object, but only documents that have a kernel have a `ClientSession`. Conversely, a console has a `ClientSession`, but since it is not backed by a file, it does not have a `Context`.

**Comment by TK-21st** [comment]
Thanks @ian-r-rose ! Been struggling with `Context` for a long time, seems like it is only needed if a widget is specified related to a document (particularly the model of the document).

Would such concepts be explicated in the future documentations? I feel like some of these concepts are really hard to understand without dedicated documentations.

**Comment by ian-r-rose** [comment]
There is actually a bit more documentation on contexts in #4453, which should be merged today. You are right, they are only needed if the widget is related to a file-backed document.

**Comment by ian-r-rose** [comment]
Closing this as answered, but feel free to ask further if something is still unclear @TK-21st
